File: problems/cpp/valid-parentheses.cc
bool isValid(string s)
{
    stack<char> brackets;
    for (auto c : s) {
        if (c == '(' | c == '{' | c == '[')
            brackets.push(c);
        else if (c == ')' && !brackets.empty() && brackets.top() == '('
                || c == '}' && !brackets.empty() && brackets.top() == '{'
                || c == ']' && !brackets.empty() && brackets.top() == '[')
            brackets.pop();
        else
            return false;
    }
    return brackets.empty();
}

File: code/python/valid-parentheses.py
# Date  : 03/25/2018
# Time  : O(n)
# Space : O(n)

class Solution:
    def isValid(self, s):
        """
        :type s: str
        :rtype: bool
        """
        stack = []
        pairs = {')' : '(', '}' : '{', ']' : '['}
        for bracket in s:
            if bracket in pairs:
                if not stack or stack[-1] != pairs[bracket]:
                    return False
                stack.pop()
            else:
                stack.append(bracket)
        return not stack
